cms/djangoapps/api/docs/decisions/001-course-apps-api.rst

```diff
@@ -0,0 +1,153 @@
+
+Each app will have the following associated metadata:
+
+- **id**: A unique identifier for the app. This can be used in the frontend to
+  look up the app's friendly title and description.
+- **enabled**: Is this app enabled for the current course.
+- **permissions**: Apps can potentially enable/disable certain operations, for
+  users based on their role. The permissions a user has can be listed here. Apps
+  may have their own permissions but at this very least this will include:
+
+    - **enable**: Can the current user enable/disable this app.
+    - **configure**: Can the current user can configure this app.
+
+  Apps may have their own custom fine-grained controls over what settings a
+  particular user can configure.
+
+  If an app doesn't have any configuration it can leave out the ``configure``
+  key and the UI will simply not show a configuration option for that app.
+- **legacy_link**: If available, this will point to the legacy link for
+  configuring the app. This can be provided as a fallback while the new UX is
+  still in development.
+
+Here is the structure that will form the basis of the API's response:
+
+.. code-block:: python
+
+    {
+        'id': 'courseapp',
+        'enabled': False,
+        'permissions': {
+            'enable': True,
+            'configure': True,
+            'edit_lti_config': False,
+        },
+        'legacy_link': 'https://studio.example.com/course_id/app-page',
+    }
+
+
+This API can be hosted at: ``/course_apps/v1/apps/{course_id}/``
+
+A ``GET`` request to this API will return an array of objects with the above
+structure. A ``PATCH`` request to the same endpoint with just the ``id`` and the
+`enabled` attribute can be used to enable/disable the app if the user has the
+requisite permissions.
+
+Each app will also have a user-friendly title and description, however this will
+be defined in the frontend since that is where internationalisation will happen.
+The frontend can look up the title and description based on the app's ``id``.
+
+This data will be provided by a plugin configuration that is part of the app.
+For old/existing apps, we can create plugin configurations that call back to
+existing code.
+
+To maintain consistency around how a course app is enabled/disabled, we can use
+Course Waffle Flags as the mechanism for enabling/disabling a course app for a
+particular course or globally. We can create a new Waffle namespace called
+"course_apps", and the course app ID can be the name for the
+``CourseWaffleFlag`` for each plugin by convention.
+
+To enumerate the list of available apps, we will list the installed plugins,
+check which plugins are available for the current course, and call each plugin
+to get its metadata. Each plugin will support checking if the plugin is
+available for a particular course. If the plugin marks itself as unavailable, it
+won't be listed in the API response.
+
+Only legacy apps will support the ``legacy_link``, it will not be a publicised
```

Comment:
Okay, when you phrase it this way, I'm down with the `legacy_link` - sounds good. :)

Reply:
Maybe I am missing something but what are `legacy apps`?

Reply:
By legacy apps, I mean existing course apps that already exist before this new system was created. These apps will need to be updated to fit into the new model of course apps. So, calculator, Wiki etc all are legacy apps. 

As for the `legacy_link`. For apps like textbooks we don't have a UI in pages and resources yet. So they need to link back to studio, and we need to provide this link in the API. The way to do that will be by supplying a `legacy_link` method. 

We will check for the presence of this method, but we don't want to expect new apps to use this part of the API. So we can mark this part of the API as depricated.

Reply:
Right, got you.